test/xbgpu/test_engine.py

```diff
@@ -327,58 +327,69 @@ async def _send_data(
+        # for a corrprod's final accumulation is not equal to its
+        # `heap_accumulation_threshold`, the accumulation index is discarded
+        # from the list (used later on). If this function attempted to receive
+        # data for an XPipeline's final, incomplete accumulation, it would
+        # result in a _spead2.Stopped error as that accumulation would never
+        # have been processed/transmitted by the XPipeline.
+        # - i.e. the test would attempt to get one more heap than the stream
+        #   contained.
+        acc_index_heap_counts: list[dict[int, int]] = []
         for corrprod_output in corrprod_outputs:
-            acc_index_counts.append(
+            acc_index_heap_counts.append(
                 dict.fromkeys(
                     [batch_index // corrprod_output.heap_accumulation_threshold for batch_index in batch_indices], 0
                 )
             )
 
-        # NOTE: The data generation generates an extra heap or several for the XPipeline
-        # with the smallest heap-accumulation-threshold - resulting in an incomplete final
-        # accumulation. Incomplete accumulations are purposefully dictated by `batch_indices`
-        # constructed by the caller.
         for i, corrprod_output in enumerate(corrprod_outputs):
             for batch_index in batch_indices:
-                adjusted_index = batch_index // corrprod_output.heap_accumulation_threshold
-                acc_index_counts[i][adjusted_index] += 1
-            # `adjusted_index` should be the last accumulation index for this corrprod
-            if acc_index_counts[i][adjusted_index] != corrprod_output.heap_accumulation_threshold:
-                acc_index_counts[i].pop(adjusted_index)
+                acc_index = batch_index // corrprod_output.heap_accumulation_threshold
+                acc_index_heap_counts[i][acc_index] += 1
+            # `acc_index` should now be the last accumulation index for this corrprod
+            if acc_index_heap_counts[i][acc_index] != corrprod_output.heap_accumulation_threshold:
+                acc_index_heap_counts[i].pop(acc_index)
 
         for batch_index in batch_indices:
             heaps = heap_factory(batch_index)
             await feng_stream.async_send_heaps(heaps, spead2.send.GroupMode.ROUND_ROBIN)
 
-        for q in mock_recv_streams:
-            q.stop()
+        for queue in mock_recv_streams:
+            queue.stop()
 
         n_baselines = n_ants * (n_ants + 1) * 2
-        device_results = np.zeros(
-            shape=(
-                len(corrprod_outputs),
-                max(len(acc_indices) for acc_indices in acc_index_counts),
-                n_channels_per_substream,
-                n_baselines,
-                COMPLEX,
-            ),
-            dtype=np.int32,
-        )
+        device_results = [
+            np.zeros(
+                shape=(
+                    len(acc_indices.keys()),  # n_accumulations for this XPipeline
```

Comment:
I was quite happy with the explicit `.keys` when iterating, since the fact that iterating a dictionary iterates the keys can be a bit non-intuitive. But `len` isn't the same thing as iteration - it just tells you how big the dictionary is (or as you've written it, how big the set of keys in the dictionary is).

Anyway, as mentioned it's just bike-shedding; feel free to leave as is if it makes more sense to you.